File: 29th/jeakim/cutting_rollcake.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
	int st, end, mid = 0;
	int top_st[10001] = {0};
	int top[10001] = {0};

	for (int i = 0; i < topping.size(); i++){
		top[topping[i]]++;
		if (top[topping[i]] == 1)
			end++;
	}
	for (int i = 0; i < topping.size(); i++){
		if(top[topping[i]] == 1)
			end--;
		top[topping[i]]--;
		top_st[topping[i]]++;
		if (top_st[topping[i]] == 1)
			st++;
		if (st == end)
			answer++;
	}
    return answer;
}

int main()
{
	vector<int> topping = {1, 2, 1, 3, 1, 4, 1, 2};
	
	cout << solution(topping) << endl;

	return (0);
}

File: 29th/jongmlee/CuttingRollcake.java
import java.util.*;

class CuttingRollcake {
    public int solution(int[] topping) {
        int answer = 0;
        Map<Integer, Integer> map = new HashMap<>();
        Set<Integer> set = new HashSet<>();
        for (int x : topping) {
            map.put(x, map.getOrDefault(x, 0) + 1);
        }
        for (int x : topping) {
            if (map.get(x) == 1) {
                map.remove(x);
            } else {
                map.put(x, map.get(x) - 1);
            }
            set.add(x);
            if (map.size() == set.size()) {
                answer++;
            }
         }
        return answer;
    }
}